### PR TITLE
CWE-352: Use of constant `state` in Oauth2 flow

### PR DESCRIPTION
**Pull request by gagliardetto**
[CWE-352: Cross-Site Request Forgery (CSRF)](https://cwe.mitre.org/data/definitions/352.html)

This query looks for cases where the oauth2 client uses a constant-value `state` string when creating the auth URL, which makes CSRF possible.

### CODE REVIEW

File: ql/src/experimental/CWE-352/ConstantOauth2State.ql [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+import go
+import DataFlow::PathGraph
+
+/*
+ * A method that creates a new URL that will send the user
+ * to the oauth2 authorization dialog of the provider.
+ */
+
```
[the remark this answers is not in the document]
Comment (smowton) [reply ql/src/experimental/CWE-352/ConstantOauth2State.ql:19]: These two should probably also be `/**` comments

[the remark this answers is not in the document]
Comment (sauyon) [reply ql/src/experimental/CWE-352/ConstantOauth2State.ql:19]: (I believe the reason for the blank lines is that the autoformatter doesn't like putting non-doc comments next to elements)

[the remark this answers is not in the document]
Comment (gagliardetto) [reply ql/src/experimental/CWE-352/ConstantOauth2State.ql:19]: Thanks for the review, @owen-mc @smowton @sauyon 

I guess I was using the wrong kind of comment notation.

**Comment by gagliardetto** [comment]
> My only reservation is how this would treat a combination of a constant value and a non-constant value. I think that currently the query would report that, but it shouldn't. Luckily this is easy to change: just make ConstantStateFlowConf extend DataFlow::Configuration instead of TaintTracking::Configuration. Please also add a test case for this situation.

Yep, found that in a test run on lgtm.com

I think the best solution would be to keep the taint analysis, but only exclude cases where the constant joins a variable.

Now I just need to figure out how to write that into the query. Any suggestions? Maybe a similar situation in another query?

**Comment by sauyon** [comment]
> I think the best solution would be to keep the taint analysis, but only exclude cases where the constant joins a variable.

It seems to me like this isn't really necessary. What patterns are you hoping to catch by using taint flow? I think it might be better to include those specifically.

File: ql/src/experimental/CWE-352/ConstantOauth2State.qhelp [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<!DOCTYPE qhelp PUBLIC
+"-//Semmle//qhelp//EN"
+"qhelp.dtd">
+<qhelp>
+    <overview>
+        <p>
+            Oauth2 clients must implement CSRF protection for the redirection URI, which is typically accomplished by including a "state" value that binds the request to
```
Comment (intrigus-lgtm) [inline ql/src/experimental/CWE-352/ConstantOauth2State.qhelp:7]:
```suggestion
            OAuth 2 clients must implement CSRF protection for the redirection URI, which is typically accomplished by including a "state" value that binds the request to
```

File: ql/src/experimental/CWE-352/ConstantOauth2State.qhelp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            the user's authenticated state. The Go Oauth2 library allows to specify a "state" value which is then included in the auth code URL, and then provided back by the remote authentication server in the redirect callback, from where it must be validated; failure to do so makes the client susceptible to an CSRF attack.
```
Comment (intrigus-lgtm) [inline ql/src/experimental/CWE-352/ConstantOauth2State.qhelp:8]:
```suggestion
            the user's authenticated state. The Go OAuth 2 library allows to specify a "state" value which is then included in the auth code URL, and then provided back by the remote authentication server in the redirect callback, from where it must be validated; failure to do so makes the client susceptible to an CSRF attack.
```

File: ql/src/experimental/CWE-352/ConstantOauth2State.ql [hunk 4]
```diff
@@ -0,0 +1,48 @@
+/**
+ * @name Use of constant `state` value in Oauth2 URL.
```
Comment (intrigus-lgtm) [inline ql/src/experimental/CWE-352/ConstantOauth2State.ql:2]:
```suggestion
 * @name Use of constant `state` value in OAuth 2 URL.
```

**Also**: Can you use backticks in the name of a query? Probably not?

File: ql/src/experimental/CWE-352/ConstantOauth2State.ql [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * @description Using a constant value for the `state` in the oauth2 URL makes the application
```
Comment (intrigus-lgtm) [inline ql/src/experimental/CWE-352/ConstantOauth2State.ql:3]:
```suggestion
 * @description Using a constant value for the `state` in the OAuth 2 URL makes the application
```
The above question also applies here.

File: ql/src/experimental/CWE-352/ConstantOauth2State.ql [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ * @tags security
+ *       external/cwe/cwe-352
+ */
+
+import go
+import DataFlow::PathGraph
+
+/*
+ * A method that creates a new URL that will send the user
+ * to the oauth2 authorization dialog of the provider.
```
Comment (intrigus-lgtm) [inline ql/src/experimental/CWE-352/ConstantOauth2State.ql:17]:
```suggestion
 * to the OAuth 2 authorization dialog of the provider.
```

File: ql/src/experimental/CWE-352/ConstantOauth2State.ql (hunk 4 above)
Comment (owen-mc) [reply ql/src/experimental/CWE-352/ConstantOauth2State.ql:2]: Re: backticks in the name of a query, you can use them. None of the queries in `codeql-go` do but some in `codeql` do.

File: ql/src/experimental/CWE-352/ConstantOauth2State.qhelp (hunk 2 above)
Comment (owen-mc) [reply ql/src/experimental/CWE-352/ConstantOauth2State.qhelp:7]: "OAuth 2.0" seems to be a bit more widespread. It's what wikipedia uses, for example, and it has more hits on google.

File: ql/src/experimental/CWE-352/ConstantOauth2State.ql (hunk 5 above)
Comment (owen-mc) [reply ql/src/experimental/CWE-352/ConstantOauth2State.ql:3]: You can use backticks in the description section.

**Comment by gagliardetto** [comment]
> It seems to me like this isn't really necessary. What patterns are you hoping to catch by using taint flow? I think it might be better to include those specifically.

I was thinking about cases where a call `fmt.Sprintf` only uses constants, and scenarios like that.

**Comment by gagliardetto** [comment]
> I was thinking about cases where a call fmt.Sprintf only uses constants, and scenarios like that.

You know what? I'll leave that out. So far haven't found a single example.

**Comment by gagliardetto** [comment]
How do I fix the stubbed dependencies?

**Comment by smowton** [comment]
@gagliardetto could you elaborate what's wrong?

**Comment by gagliardetto** [comment]
This query tests depend on `golang.org/x/oauth2` package.

I need to somehow add this dependency without completely revolutionizing the go.mod file of `codeql-go` and without using the official vendoring mechanism of `go mod` (need to use `depstubber` instead).

I'm obviously doing something wrong with those two elements, but I'm not sure what.

**Comment by gagliardetto** [comment]
That's the reason why the tests fails, BTW.

**Comment by owen-mc** [comment]
@gagliardetto I believe it is failing because it doesn't have a `go.mod` file in the test folder. I made a file of that name with the following contents and then the tests passed.
```
module custom-queries-go-tests/constant-oauth2-state

go 1.14

require golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
```

**Comment by gagliardetto** [comment]
Thanks everyone!

**Pull request merged**